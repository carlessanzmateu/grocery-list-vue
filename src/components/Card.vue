<template>
  <div class="card w-full relative rounded-sm">
    <span :class="[getFavoriteClasses]" @click="toggleFavorite"></span>
    <img
      class="image w-full bg-cover rounded-t-sm"
      :alt="item.getDescription()"
      :src="item.getImage()"
    />
    <div class="card-information p-2">
      <h3 class="title truncate w-full"><strong>{{ item.getName() }}</strong></h3>
      <p class="description truncate w-full mb-2" :title="item.getDescription()">
        {{ item.getDescription() }}
      </p>
      <div class="buy-info flex justify-between">
        <span class="products-left pt-2">Stock: {{ item.getStockLeft() }}</span>
        <span class="products-left pt-2">Price: {{ item.getPrice() }}$</span>
      </div>
      <a
          class="add-btn rounded-sm cursor-pointer text-center p-1"
          @click="buyItem"
        >
        Comprar
      </a>
    </div>
  </div>
</template>

<script>
export default {
  name: 'Card',
  props: {
    item: {
      type: Object,
      required: true
    }
  },
  computed: {
    getFavoriteClasses () {
      return {
        favorite: true,
        filled: this.item.getIsFavorite()
      }
    }
  },
  methods: {
    toggleFavorite () {
      this.$emit('toggle-favorite', this.item)
    },
    buyItem () {
      this.$emit('buy-item', this.item)
    }
  }
}
</script>

<style lang="scss" scoped>
.card {
  max-width: 250px;
  height: 365px;
  background-color: var(--white-500);
  box-shadow: 0 1px 3px rgba(0,0,0,.12), 0 1px 2px rgba(0,0,0,.24);
  transition: all 0.3s cubic-bezier(.25, .8, .25, 1);

  &:hover {
    box-shadow: 0 14px 28px rgba(0,0,0,.25), 0 10px 10px rgba(0,0,0,.22)
  }

  .favorite:before {
    content: '\2665';
    color: var(--white-500);
    font-size: 40px;
    position: absolute;
    right: 15px;
    cursor: pointer;

  }

  .filled:before {
    color: var(--red-500);
  }

  .image {
    height: 60%;
  }

  .card-information {
    .title {
      font-size: 19px;
    }

    .description {
      font-size: 17px;
    }

    .add-btn {
      height: 40px;
      color: var(--white-500);
      background-color: var(--blue-500);
      line-height: 37px;

      &:hover {
        background-color: var(--violet-500);
      }
    }
  }

  @media (max-width: 768px) {
    max-width: 350px;
  }

  @media (max-width: 768px) {
    max-width: 250px;
  }
}
</style>
