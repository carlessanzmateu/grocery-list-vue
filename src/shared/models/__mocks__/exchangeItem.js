const mockExchangeItem = {
  id: '1',
  // eslint-disable-next-line
  image_url: 'mock.png',
  stock: 1,
  productName: 'some name',
  price: 10,
  productDescription: 'Some description',
  favorite: 0
}
