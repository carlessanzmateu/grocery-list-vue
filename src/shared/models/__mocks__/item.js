class MockItem {
  constructor() {
    this.$id = '1'
    this.$image = 'mock.png'
    this.$stockReadyToBuy = 0
    this.$stock = 1
    this.$name = 'some name'
    this.$price = 10
    this.$description = 'Some description'
    this.$favorite = false
  }

  getId () {
    return this.$id
  }

  getImage () {
    return this.$image
  }

  initSelectedStock () {
    this.$stockReadyToBuy = 0
  }

  getSelectedStock () {
    return this.$stockReadyToBuy
  }

  getStock () {
    return this.$stock
  }

  getStockLeft () {
    return this.getStock() - this.getSelectedStock()
  }

  canIncreaseSelectedStock () {
    return this.getSelectedStock() < this.getStock()
  }

  canDecreaseSelectedStock () {
    return this.getSelectedStock() > 0
  }

  increaseSelectedStock () {
    if (!this.canIncreaseSelectedStock()) {
      return
    }

    this.$stockReadyToBuy++
  }

  decreaseSelectedStock () {
    if (!this.canDecreaseSelectedStock()) {
      return
    }

    this.$stockReadyToBuy--
  }

  getName () {
    return this.$name
  }

  getPrice () {
    return this.$price
  }

  getItemTotalPrice () {
    return this.getSelectedStock() * this.getPrice()
  }

  getDescription () {
    return this.$description
  }

  buildFavorite (favorite) {
    this.$favorite = (favorite === '1')
  }

  getIsFavorite () {
    return this.$favorite
  }

  setAsFavorite () {
    this.$favorite = true
  }

  removeFromFavorite () {
    this.$favorite = false
  }

  toggleFavorite () {
    this.$favorite = !this.$favorite
  }
}

export { MockItem }
