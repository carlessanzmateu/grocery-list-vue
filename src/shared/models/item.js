export default class Item {
  $id
  $image
  $stockReadyToBuy
  $stock
  $name
  $price
  $description
  $favorite

  constructor (rawItem) {
    this.$id = rawItem.id
    this.$image = rawItem.image_url
    this.$stock = rawItem.stock
    this.$name = rawItem.productName
    this.$price = rawItem.price
    this.$description = rawItem.productDescription
    this.buildFavorite(rawItem.favorite)
    this.initSelectedStock()
  }

  getId () {
    return this.$id
  }

  getImage () {
    return this.$image
  }

  initSelectedStock () {
    this.$stockReadyToBuy = 0
  }

  getSelectedStock () {
    return this.$stockReadyToBuy
  }

  // setSelectedStock (selectedStock) {
  //   this.$stockReadyToBuy = selectedStock
  // }

  getStock () {
    return this.$stock
  }

  setStock (stock) {
    this.$stock = stock
  }

  getStockLeft () {
    return this.getStock() - this.getSelectedStock()
  }

  canIncreaseSelectedStock () {
    return this.getSelectedStock() < this.getStock()
  }

  canDecreaseSelectedStock () {
    return this.getSelectedStock() > 0
  }

  increaseSelectedStock () {
    if (!this.canIncreaseSelectedStock()) {
      return
    }

    this.$stockReadyToBuy++
  }

  decreaseSelectedStock () {
    if (!this.canDecreaseSelectedStock()) {
      return
    }

    this.$stockReadyToBuy--
  }

  getName () {
    return this.$name
  }

  getPrice () {
    return this.$price
  }

  getItemTotalPrice () {
    return this.getSelectedStock() * this.getPrice()
  }

  getDescription () {
    return this.$description
  }

  buildFavorite (favorite) {
    this.$favorite = (favorite === '1')
  }

  getIsFavorite () {
    return this.$favorite
  }

  setAsFavorite () {
    this.$favorite = true
  }

  removeFromFavorite () {
    this.$favorite = false
  }

  toggleFavorite () {
    this.$favorite = !this.$favorite
  }
}
