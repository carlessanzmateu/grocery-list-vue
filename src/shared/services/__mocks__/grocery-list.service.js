export default {
  getAllGroceries: jest.fn(),
  getAllFavorites: jest.fn(),
  updateGrocery: jest.fn()
}
