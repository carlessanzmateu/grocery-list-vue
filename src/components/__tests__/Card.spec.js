import { shallowMount } from '@vue/test-utils'
import Card from '@/components/Card.vue'
import { MockItem } from '@/shared/models/__mocks__/item'

describe('Card', () => {
  let wrapper
  function createWrapper() {
    return shallowMount(Card, {
      propsData: {
        item: new MockItem()
      }
    })
  }

  it('should create', () => {
    wrapper = createWrapper()
    expect(wrapper.isVueInstance()).toBe(true)
  })

  it('should match html', () => {
    wrapper = createWrapper()
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should have toggleFavorite method', () => {
    wrapper = createWrapper()

    expect(wrapper.vm.toggleFavorite).toBeDefined()
  })

  it('should emit toggle-favorite event when call toggleFavorite method', () => {
    wrapper = createWrapper()

    wrapper.vm.toggleFavorite()

    expect(wrapper.emitted()['toggle-favorite'])
  })

  it('should have buyItem method', () => {
    wrapper = createWrapper()

    expect(wrapper.vm.buyItem).toBeDefined()
  })

  it('should emit buy-item event when call buyItem method', () => {
    wrapper = createWrapper()

    wrapper.vm.buyItem()

    expect(wrapper.emitted()['buy-item'])
  })
})
