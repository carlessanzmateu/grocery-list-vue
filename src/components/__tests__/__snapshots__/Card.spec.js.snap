// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match html 1`] = `
<div
  class="card w-full relative rounded-sm"
>
  <span
    class="favorite"
  />
   
  <img
    alt="Some description"
    class="image w-full bg-cover rounded-t-sm"
    src="mock.png"
  />
   
  <div
    class="card-information p-2"
  >
    <h3
      class="title truncate w-full"
    >
      <strong>
        some name
      </strong>
    </h3>
     
    <p
      class="description truncate w-full mb-2"
      title="Some description"
    >
      
      Some description
    
    </p>
     
    <div
      class="buy-info flex justify-between"
    >
      <span
        class="products-left pt-2"
      >
        Stock: 1
      </span>
       
      <span
        class="products-left pt-2"
      >
        Price: 10$
      </span>
    </div>
     
    <a
      class="add-btn rounded-sm cursor-pointer text-center p-1"
    >
      
      Comprar
    
    </a>
  </div>
</div>
`;
